draft-ietf-httpbis-priority.md

```diff
@@ -369,70 +380,82 @@ Stream ID:
- +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+HTTP/3 PRIORITY_UPDATE Frame {
+  Type (i) = 0xF0700..0xF0701,
+  Length (i),
+  Prioritized Element ID (i),
+  Priority Field Value (..),
+}
 ~~~
-{: #fig-h3-reprioritization-frame title="HTTP/3 PRIORITY_UPDATE Frame Payload"}
+{: #fig-h3-reprioritization-frame title="HTTP/3 PRIORITY_UPDATE Frame"}
 
 The PRIORITY_UPDATE frame payload has the following fields:
 
-T (Prioritized Element Type):
-: A one-bit field indicating the type of element
-being prioritized. A value of 0 indicates a reprioritization for a Request
-Stream, so the Prioritized Element ID is interpreted as a Stream ID. A
-value of 1 indicates a reprioritization for a Push stream, so the Prioritized
-Element ID is interpreted as a Push ID.
-
-Empty:
-: A seven-bit field that has no semantic value.
-
 Prioritized Element ID:
 : The stream ID or push ID that is the target of the priority update.
 
 Priority Field Value:
 : The priority update value in ASCII text, encoded using Structured Headers.
 
-The HTTP/3 PRIORITY_UPDATE frame MUST NOT be sent with an invalid identifier,
-including before the request stream has been opened or before a promised
-request has been received.  If a server receives a PRIORITY_UPDATE specifying
-a push ID that has not been promised, it SHOULD be treated as a connection
-error of type H3_ID_ERROR.
-
-Because the HTTP/3 PRIORITY_UPDATE frame is sent on the control stream and
-there are no ordering guarantees between streams, a client that reprioritizes
-a request before receiving the response data might cause the server to receive
-a PRIORITY_UPDATE for an unknown request. If the request stream ID is within
-bidirectional stream limits, the PRIORITY_UPDATE frame SHOULD be buffered
-until the stream is opened and applied immediately after the request message
-has been processed. Holding PRIORITY_UPDATES consumes extra state on the peer,
-although the size of the state is bounded by bidirectional stream limits. There
-is no bound on the number of PRIORITY_UPDATES that can be sent, so an
-endpoint SHOULD store only the most recently received frame.
+The request-stream variant of PRIORITY_UPDATE (type=0xF0700) MUST reference a
+request stream. If a server receives a PRIORITY_UPDATE (type=0xF0700) for a
+Stream ID that is not a request stream, this MUST be treated as a connection
+error of type H3_ID_ERROR. PRIORITY_UPDATE (type=0xF0700) MAY be sent before the
+request stream that it references has been created. The Stream ID MUST be within
+the client-initiated bidirectional stream limit. If a server receives a
+PRIORITY_UPDATE (type=0xF0700) with a Stream ID that is beyond the stream
+limits, this SHOULD be treated as a connection error of type H3_ID_ERROR.
+
+Request-stream variant PRIORITY_UPDATEs (type=0xF0700) received before the
+request or response has started SHOULD be buffered until the stream is opened
+and applied immediately after the request message has been processed. Holding
```

Comment:
This is about existing text, and we can defer the decision, but I think it might be better to say "SHOULD be buffered and be applied once the stream is opened."

The problem with current text is that it does deal with servers sending responses _while_ they receive requests (i.e. early response).

Reply:
Agreed that is better.

Reply:
At first I agrred but I've thought about a downside. So I've created https://github.com/httpwg/http-extensions/issues/1260 where we can discuss separately.